R/scale-.r

```diff
@@ -490,184 +792,115 @@ ScaleDiscrete <- ggproto("ScaleDiscrete", Scale,
   }
 
-  ggproto(NULL, super,
-    call = match.call(),
-
-    aesthetics = aesthetics,
-    scale_name = scale_name,
-    palette = palette,
-
-    range = continuous_range(),
+  ggproto(NULL, ViewScale,
+    scale = scale,
+    aesthetics = scale$aesthetics,
+    name = scale$name,
+    scale_is_discrete = scale$is_discrete(),
     limits = limits,
-    trans = trans,
-    na.value = na.value,
-    expand = expand,
-    rescaler = rescaler,  # Used by diverging and n colour gradients
-    oob = oob,
-
-    name = name,
+    continuous_range = continuous_range,
     breaks = breaks,
-    minor_breaks = minor_breaks,
-
-    labels = labels,
-    guide = guide,
-    position = position
+    labels = scale$get_labels(breaks),
+    minor_breaks = minor_breaks
   )
 }
 
-#' Discrete scale constructor.
-#'
-#' @export
-#' @inheritParams continuous_scale
-#' @param palette A palette function that when called with a single integer
-#'   argument (the number of levels in the scale) returns the values that
-#'   they should take.
-#' @param breaks One of:
-#'   - `NULL` for no breaks
-#'   - `waiver()` for the default breaks computed by the
-#'     transformation object
-#'   - A character vector of breaks
-#'   - A function that takes the limits as input and returns breaks
-#'     as output
-#' @param limits A character vector that defines possible values of the scale
-#'   and their order.
-#' @param drop Should unused factor levels be omitted from the scale?
-#'    The default, `TRUE`, uses the levels that appear in the data;
-#'    `FALSE` uses all the levels in the factor.
-#' @param na.translate Unlike continuous scales, discrete scales can easily show
-#'   missing values, and do so by default. If you want to remove missing values
-#'   from a discrete scale, specify `na.translate = FALSE`.
-#' @param na.value If `na.translate = TRUE`, what value aesthetic
-#'   value should missing be displayed as? Does not apply to position scales
-#'   where `NA` is always placed at the far right.
-#' @keywords internal
-discrete_scale <- function(aesthetics, scale_name, palette, name = waiver(),
-  breaks = waiver(), labels = waiver(), limits = NULL, expand = waiver(),
-  na.translate = TRUE, na.value = NA, drop = TRUE,
-  guide = "legend", position = "left", super = ScaleDiscrete) {
+view_scale_empty <- function() {
+  ggproto(
+    NULL, ViewScale,
+    is_empty = function() TRUE,
+    is_discrete = function() NA,
+    dimension = function() c(0, 1),
+    get_limits = function() c(0, 1),
+    get_breaks = function() NULL,
+    get_breaks_minor = function() NULL,
+    get_labels = function() NULL,
+    rescale = function(x) stop("Not implemented", call. = FALSE),
+    map = function(x) stop("Not implemented", call. = FALSE),
+    make_title = function(title) title,
+    break_positions = function() NULL,
+    break_positions_minor = function() NULL
+  )
+}
 
-  aesthetics <- standardise_aes_names(aesthetics)
+ViewScale <- ggproto("ViewScale", NULL,
```

Comment:
I think the code is a bit easier to follow if the parent class is defined first.